### PR TITLE
Rethrow EEI exceptions in WAVM

### PR DESCRIPTION
**Pull request by poemm**
Passes the following tests `useGas`, `getCallDataSize`, `callDataCopy`, `callDataCopy256`, `revert`, `returnPredefinedData`, `revertVeryLong`, and `returnVeryLong`.

Each host function which can throw a hera exception has a try-catch to catch and record this exception, trigger a WAVM trap so that it can clean up. Once WAVM returns, we re-throw the original hera exception.

Don't know if this is what you wanted. If not, I will close.

### CODE REVIEW

File: src/wavm.cpp [hunk 1]
```diff
@@ -193,9 +227,49 @@ ExecutionResult WavmEngine::execute(
...
+
+  // execute the contract
   ExecutionResult result = internalExecute(context, code, state_code, msg, meterInterfaceGas);
-  // And clean up mess left by this run.
+
+  // clean up this run, this is done here after leaving the scope of internalExecute()
   Runtime::collectGarbage();
+
+  // re-throw exception if there was one
+  if (wavm_host_module::exceptionIdx != wavm_host_module::HeraExceptionsEnum::None){
```
Comment (jakelang) [inline src/wavm.cpp]: maybe simplify using a helper?

File: src/wavm.cpp [hunk 2]
```diff
@@ -193,9 +227,49 @@ ExecutionResult WavmEngine::execute(
   evmc_message const& msg,
   bool meterInterfaceGas
 ) {
+
+  // clear stuff from previous run
+  wavm_host_module::exceptionIdx = wavm_host_module::HeraExceptionsEnum::None;
+  wavm_host_module::exceptionString = "";
+  // hope and prayer
+  Runtime::collectGarbage();
```
Comment (jakelang) [inline src/wavm.cpp]: why do we collect garbage redundantly here? (other call at L241)

File: src/wavm.cpp [hunk 3]
```diff
@@ -143,13 +157,33 @@ namespace wavm_host_module {
...
   }
 
 
   DEFINE_INTRINSIC_FUNCTION(ethereum, "revert", void, revert, U32 dataOffset, U32 length)
   {
-    interface.top()->eeiRevert(dataOffset, length);
+    try {
+      interface.top()->eeiRevert(dataOffset, length);
+    } catch (InvalidMemoryAccess const& e) {
+      HERA_DEBUG<<"caught Hera's InvalidMemoryAccess\n";
```
[the remark this answers is not in the document]
Comment (poemm) [reply src/wavm.cpp]: I agree that this debug message could be more descriptive. How about outputting the original exception message `HERA_DEBUG << e.what()`? Or `HERA_DEBUG << "caught Hera's exception: "<< e.what() << ". Will exit WAVM invokation now." `? Or just removing this debug message altogether?

File: src/wavm.cpp (hunk 2 above)
Comment (poemm) [reply src/wavm.cpp]: Yes, this is a "hope and prayer" that it will clean up something. Removing this is OK if you insist.

File: src/wavm.cpp (hunk 1 above)
Comment (poemm) [reply src/wavm.cpp]: Could do it if you insist. But `hera_execute()` does a similar long `switch-case` inside of an `if` without a helper function.

File: src/wavm.cpp [hunk 4]
```diff
@@ -143,13 +157,33 @@ namespace wavm_host_module {
 
   DEFINE_INTRINSIC_FUNCTION(ethereum, "finish", void, finish, U32 dataOffset, U32 length)
   {
-    interface.top()->eeiFinish(dataOffset, length);
```
Comment (axic) [inline src/wavm.cpp]: This also throws `EndExecution` which isn't a failure.

File: src/wavm.cpp [hunk 5]
```diff
@@ -143,13 +157,33 @@ namespace wavm_host_module {
...
+      throwException(Runtime::Exception::calledUnimplementedIntrinsicType); //or maybe reachedUnreachableType
+    }
   }
 
 
   DEFINE_INTRINSIC_FUNCTION(ethereum, "revert", void, revert, U32 dataOffset, U32 length)
   {
-    interface.top()->eeiRevert(dataOffset, length);
+    try {
+      interface.top()->eeiRevert(dataOffset, length);
```
Comment (axic) [inline src/wavm.cpp]: This also throws `EndExecution` with revert set.

**Comment by axic** [comment]
All the EEI methods can throw multiple exceptions (potentially 7 different ones). Given we implement ~10 EEI methods in this file and each exception catching is around 10 lines of code if we do not generalise this then this will take ~700 lines of code.

**Comment by poemm** [comment]
Thank you for the tip, `std::current_exception` would be more concise and maintainable.

For this to work, we need a `std::exception_ptr` in the scope of the host functions, I see nowhere else to declare this ptr but next to `stack<WavmEthereumInterface*>`. This ptr must be nulled after we rethrow it (so that we don't accidentally reuse this exception in different calls), I see nowhere else to null it but at the bottom of each `call*` host function, and also at the beginning of `WavmEngine::execute()`. Not sure if there will be any ambiguities or edge cases.

**Comment by poemm** [comment]
Granted @chfast 's request. All tests in the original post are still passing.

**Comment by axic** [comment]
> testeth 1.4.0rc7

This PR is using a very old Hera. It needs to be rebased to make tests pass. Also there are some typos the linter is picking up.

**Comment by poemm** [comment]
I fetched upstream, then did `git rebase upstream/master wavm_rethrow_exceptions`, then fixed typos, then force pushed.

File: src/wavm.cpp [hunk 6]
```diff
@@ -71,10 +77,16 @@ unique_ptr<WasmEngine> WavmEngine::create()
...
 
+
 namespace wavm_host_module {
+
   // first the ethereum interface(s), the top of the stack is used in host functions
   stack<WavmEthereumInterface*> interface;
 
+  // set up for re-throwing exceptions, this is not clean, but Hera and WAVM are incompatible without some awkwardness 
+  std::exception_ptr my_exception_ptr;
+  bool VMTrapFlag;
```
Comment (chfast) [inline src/wavm.cpp]: Init to `false`.

File: src/wavm.cpp [hunk 7]
```diff
@@ -57,6 +59,10 @@ class WavmEthereumInterface : public EthereumInterface {
     m_wasmMemory = _wasmMemory;
   }
 
+  void nullWasmMemory() {
```
Comment (axic) [inline src/wavm.cpp]: Instead of this you could use `setWasmMemory(nullptr)`.

File: src/wavm.cpp [hunk 8]
```diff
@@ -120,6 +136,9 @@ namespace wavm_host_module {
   DEFINE_INTRINSIC_FUNCTION(ethereum, "storageStore", void, storageStore, U32 pathOffset, U32 valueOffset)
   {
     interface.top()->eeiStorageStore(pathOffset, valueOffset);
+    // print stuff for debugging
+    //interface.top()->debugPrintMem(1, valueOffset, 32);
+    //interface.top()->debugPrintStorage(1, pathOffset);
```
Comment (axic) [inline src/wavm.cpp]: Need to be removed before merging.

**Comment by poemm** [comment]
@chfast I upgraded wavm to the version used in https://github.com/ewasm/hera/pull/422 and implemented exception rethrow in all host functions. The tests listed above pass, but there is still a memory leak. Can we upgrade to the newest version of wavm to see if the memory leak remains? I tried upgrading, but there were build errors -- the cmake stuff may be less difficult for you.

**Comment by axic** [comment]
WAVM has a lot of commits lately, including [this documentation on exceptions](https://github.com/WAVM/WAVM/commit/c1d2ba945edcc9d87e9fc0044777ab1850a1c2a6).

Maybe exception handled is better now?

**Comment by axic** [comment]
I think this actually isn't needed. The problem in #425 wasn't due to the fact that exceptions are not passed through, but that the `interface` stack was not freed if an exception occurred and thus in a case of a contract calling another after returning from a call it will messed up.
